File: scala/day1/compass.scala
class Compass {
  val directions = List("N", "E", "S", "W")
  var bearing = 0
  println("Initial bearing: " + directions(bearing))

  def direction() = directions(bearing)

  def inform(turnDirection: String) = {
    println(s"Turning $turnDirection. Now bearing ${direction()}")
  }

  def turnRight() = {
    bearing = (bearing + 1) % directions.size
    inform("right")
  }
  def turnLeft() = {
    bearing = (bearing + (directions.size - 1)) % directions.size
    inform("left")
  }
}

@main def main() = {
  val compass = new Compass
  compass.turnRight()
  compass.turnRight()
  compass.turnRight()
  compass.turnRight()
  compass.turnRight()
  compass.turnLeft()
}

File: clojure/day2/compass.clj
(ns compass
  (:gen-class))
(println "----------------")

(defprotocol Compass
  (direction [c])
  (left [c])
  (right [c]))

(def directions [:north :east :south :west])

(defn turn [base amount] (rem (+ base amount) (count directions)))

(defrecord SimpleCompass [bearing]
  Compass
  (direction [_] (directions bearing))
  (left [_] (SimpleCompass. (turn bearing 3)))
  (right [_] (SimpleCompass. (turn bearing 1)))
  (toString [this] (str "SimpleCompass pointing at " (direction this))))


(def c (SimpleCompass. 0))
(println c)
(println (left c))
(println (right c))

(println (direction c))
(println (direction (left c)))
(println (direction (right c)))

(println (str c))

(println "----------------")

(def d (SimpleCompass. 2))
(println d)
(println (str d))
(println (direction d))
(println (:bearing d))
(println (directions (:bearing d)))

(println "----------------")

(def e (SimpleCompass. 2))
(println (str d))
(println (str e))
(println "=?" (= d e))
